src/main/java/com/redhat/foreman/ForemanComputerListener.java

```diff
@@ -30,4 +33,24 @@ public void onLaunchFailure(Computer c, TaskListener taskListener) throws IOExce
         }
     }
 
+    @Override
+    @SuppressFBWarnings(value = "BC_VACUOUS_INSTANCEOF")
+    public void preLaunch(Computer c, TaskListener taskListener) throws IOException, InterruptedException {
+        try {
+            super.preLaunch(c, taskListener);
+        } catch (Exception e) {
+            LOGGER.log(Level.SEVERE, "Uncaught unexpected exception occurred while calling super.onLaunchFailed(): ", e);
+        }
+        if (c instanceof ForemanComputer) {
+            Node node = c.getNode();
+            if (node instanceof ForemanSharedNode) {
+               ForemanSharedNodeCloud foremanCloud =
+                       ForemanSharedNodeCloud.getByName(((ForemanSharedNode) node).getCloudName());
+               if (foremanCloud == null || !foremanCloud.isOperational()) {
+                   throw new AbortException("This is a leaked ForemanSharedNode after Jenkins restart!");
```

Comment:
Definitely, BUT we've implemented `EphemeralNode` in the `ForemanSharedNode`, so this hook `ComputerListener.preLaunch()` isn't called after Jenkins init is completed and *before* we allow to provision a new slave...